### DISCUSSION TITLE
JavaScript package for CUE using wasm

### DISCUSSION THREAD

**Discussion by codewithcheese** [cuelang/cue#952] [Show and tell]
Hi CUEbies

I have created a npm packages for using CUE in node.js and browsers: https://www.npmjs.com/package/cuelang-js

It compiles the CUE cli to wasm and exports a function for calling commands along with providing arguments and flags. In node.js it reads from the local file system, in browser CUE reads and writes from an in-memory file system.

The CUE wasm is 26MB uncompressed using gzip its down to 4MB still a bit large for websites.

We at [balena](https://www.balena.io/) are starting to use CUE in many places. One of them is to replace JSON Schema for validation.

If you make use of the package feel free to drop me a note here, happy to accept issues and PRs on the repo.

**Comment by myitcv** [comment]
@codewithcheese out of interest, how do you plan to version this? Because there are effectively two relevant versions that compose into the output: the Go version you used to build the WASM artefact, and the CUE version. On top of that you've got your "version" which effectively represents the progression of API you are establishing in your package. I just wonder how this scales as there are more Go versions, more CUE versions... and your API could change too?

**Reply by codewithcheese** [reply]
@myitcv that's a good question. In retrospect I now would have preferred to mirror the CUE version so that the CUE version and JS package version are in sync. In which case breaking changes to the JS API could also be carried out between breaking versions of CUE. Currently the JS API is extremely simple being just a function that takes a command string, arguments and flags. I can't see that changing much. But I have considered adding helper functions for each of the commands which would then see a lot more flux as commands are modified.

As it stands cuelang-js is already passed version 1 and I plan to maintain the compatibility table which shows which CUE version is included in which release. As far a Go version I had not given it much thought, I consider the WASM like a binary and i'm usually not too concerned which version of a language was used to compile a binary.

What do you think?

**Reply by myitcv** [reply]
> I can't see that changing much

Famous last words :)



> As it stands cuelang-js is already passed version 1 and I plan to maintain the compatibility table which shows which CUE version is included in which release

I would perhaps look to how GopherJS are handling a not dissimilar situation:

https://github.com/gopherjs/gopherjs



> As far a Go version I had not given it much thought, I consider the WASM like a binary and i'm usually not too concerned which version of a language was used to compile a binary.

The Go version will exclusively define, for a given version of CUE, the resulting WASM binary size, as well as the efficiency of code etc. So I would say this is a significant detail. 

Honestly I don't have a good answer, other than to suggest looking at how GopherJS has tried to tackle this problem recently.

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/952.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
